File: 118.pascals-triangle.java
/*
 * @lc app=leetcode id=118 lang=java
 *
 * [118] Pascal's Triangle
 */

// @lc code=start

import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<List<Integer>> generate(int numRows) {
        if (numRows == 1) {
            return List.of(List.of(1));
        }
        if (numRows == 2) {
            return List.of(
                    List.of(1), List.of(1, 1));
        }
        List<List<Integer>> above = new ArrayList<>(generate(numRows - 1));
        List<Integer> prev = above.get(above.size() - 1);
        List<Integer> lst = new ArrayList<>();
        lst.add(1);
        for (int ix = 0; ix != numRows - 2; ++ix) {
            lst.add(prev.get(ix) + prev.get(ix + 1));
        }
        lst.add(1);
        above.add(lst);
        return above;
    }
}
// @lc code=end

File: 118.pascals-triangle.cpp
/*
 * @lc app=leetcode id=118 lang=cpp
 *
 * [118] Pascal's Triangle
 */

// @lc code=start
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> generate(int numRows) {
        vector<vector<int>> ret;
        for (auto row = 0ull; row != numRows; ++row) {
            vector<int> curr(row + 1, 1);
            for (auto col = 1ull; col < row; ++col) {
                curr[col] = ret[row - 1][col - 1] + ret[row - 1][col];
            }
            ret.push_back(std::move(curr));
        }
        return ret;
    }
};
// @lc code=end
